---
title: "DZ TDP43 concentration dependent analysis"
author: "Matteo Zanovello"
date: "2022-01-10"
output:
---

```{r setup, include = False}
knitr::opts_chunk$set(warning = FALSE, message = FALSE)
if (!require("pacman")) {install.packages("pacman")}
if (!requireNamespace("BiocManager", quietly = TRUE)) {install.packages("BiocManager")}
pacman::p_load(tidyverse, data.table, janitor, ggpubr, ggrepel, magrittr, biomaRt, tximport,
               devtools, readxl, DESeq2, edgeR, annotables, rtracklayer, ensembldb,
               PCAtools, clusterProfiler, "org.Hs.eg.db", enrichplot, pcaExplorer, topGO, UpSetR,
               ggtranscript, "TxDb.Hsapiens.UCSC.hg38.knownGene", GenomicRanges)
            
#source(here::here('scripts','create_feature_count_table.R'))
#source(here::here('scripts','run_standard_deseq.R'))
source(here::here('scripts','salmon_deseq2.R'))
source(here::here('scripts','make_volcano_plot.R'))
source(here::here('scripts','splicing_volcanoplot.R'))
#source(here::here('scripts','trial_graph_concentration.R'))
#source(here::here('scripts','run_de_irfinder.R'))
#source(here::here('scripts','ggtranscript.R'))
```
###levels in qpcr
```{r}
#data from qpcr
levels_dz_qpcr <- fread("/Users/matteozanovello/Documents/phd/research_lines/tdp-43 curves/tdp curves dz/2022-05-04_145923_CURVES_DZ1.csv")
levels_dz_qpcr %>%
  filter(experiment == "new_new") %>%
  ggplot(aes(x = as.character(sample), y = 100*value)) +
  geom_point() +
  geom_boxplot() +
  xlab(expression(paste('Doxycycline concentration (', mu,'g/mL)'))) +
  ylab('TDP-43 levels compared to untreated (%)') +
  #facet_wrap(facets = vars(experiment)) +
  theme_classic() +
  theme(text = element_text(size = 24))
#ggsave(file = "~/Desktop/qpcr_dz_curves.png", width = 8)

levels_sy5y_qpcr <- fread("/Users/matteozanovello/Documents/phd/research_lines/tdp-43 curves/unc13a_b_tdp43_curves_nature_paper.csv")
levels_sy5y_qpcr %>%
  #filter(experiment == "new") %>%
  ggplot(aes(x = as.character(Treatment), y = Value)) +
  geom_point() +
  geom_boxplot() +
  xlab(expression(paste('Doxycycline concentration (', mu,'g/mL)'))) +
  ylab('TDP-43 levels compared to untreated (%)') +
  #facet_wrap(facets = vars(experiment)) +
  theme_classic() +
  theme(text = element_text(size = 24))
#ggsave(file = "~/Desktop/qpcr_sy_curves.png", width = 8)
#levels_dz_qpcr %>%
#  group_by(experiment,sample) %>%
#  summarise_at(vars(value),list(name = mean))
```

this organizes all the data from the many .bam files and creates one data frame you can work with,
then runs the make_deseq_dfs() function and run_standard_deseq() and eventually make graphs
```{r}
sample_dir_dz <- "~/Documents/GitHub/tdp43_concentration/data/salmon_quant_dz/"
metadata_dir_dz <- fread("~/Documents/GitHub/tdp43_concentration/data/metadata_dz_curves.csv")
levels_dz = c(0.02, 0.03, 0.04, 0.05, 0.075, 0.1, 1)

for (j in levels_dz) {
  salmon_deseq2(j, sample_dir_dz, metadata_dir_dz)
  bbb <- label_significant(results_table, log2FoldCut = 2, log10padj = 5) + annotate("text", x=0.9, y=0.1, label=j)
  #plot(bbb)
  assign(paste0("normed_counts_dz_curves_",j), normed_counts)
  assign(paste0("results_table_dz_curves_",j), results_table)
}

AppendMe <- function(dfNames) {
  do.call(rbind, lapply(dfNames, function(x) {
    cbind(get(x), source = x)
  }))
}

#funnel plot with differential gene expression
list_de_dz <- paste0("results_table_dz_curves_", levels_dz)
big_delta_de_dz <- AppendMe(list_de_dz) 
big_delta_de_dz$source <- factor(big_delta_de_dz$source, levels = list_de_dz, labels = levels_dz)
list_gene <- c("UNC13A", "STMN2"#, "AARS", #"UNC13B", 
               #"CELF5", "ELAVL3", "INSR", "ATG4B", #"KCNQ2", #"RANBP1", 
               #"NUP188","SYT7", #"WASL",
                #"RAB1GAP", #"AGRN","PFKP", 
                 #"HDGFL2","ARHGAP32", "CEP290", "PRELID3A", "KALRN",
                 #"CAMK2B", "ADCY1", "TRAPPC12", "EPB41L4A", "IQCE", "SEMA4D"
                 #"CYFIP2", "SYNE1"
               )

plot_de_dz <- big_delta_de_dz %>%
  mutate(color_gene_name = ifelse(log2FoldChange > 0, "up","down")) %>%
  mutate(alpha_gene_name = ifelse(padj < 0.05, "sig", "non_sig")) %>%
  mutate(gene_label = case_when(symbol %in% list_gene & alpha_gene_name == "sig" ~ symbol, T ~ ""))
plot_de_dz_meta <- as.data.frame(table(plot_de_dz$color_gene_name, plot_de_dz$alpha_gene_name, plot_de_dz$source))

plot_de_dz %>%
  ggplot(aes(x = log2FoldChange, y = source)) +
  geom_point(aes(color = color_gene_name, alpha = alpha_gene_name),
             size = 0.2,
             position = position_jitter(width = 0, height = 0.4),
             show.legend = F) +
  geom_text(data = plot_de_dz_meta[plot_de_dz_meta$Var2 == "sig",], 
            aes(x = ifelse(Var1 == "down", -8, 8), y = Var3, label = Freq, color = Var1),
            size = 5,
            show.legend = F) +
  geom_vline(xintercept = 0) +
  geom_text_repel(aes(label = gene_label), max.overlaps = Inf) +
  #coord_flip() +
  scale_color_brewer(palette = "Set1") +
  scale_alpha_manual(values = c(0.2, 1)) +
  xlab('Log2 Fold Change compared to untreated') +
  ylab('Doxycycline concentration (ug/mL)') +
  xlim(c(-10,10)) +
  theme_classic() 


salmon_deseq2(levels_dz, sample_dir_dz, metadata_dir_dz)
assign("normed_counts_dz_curves", normed_counts)

#write.csv(normed_counts_dz_curves, file = "~/Documents/GitHub/tdp43_concentration/results/normed_counts_dz/normed_counts_dz_all.csv") #"normed_counts_dz_all.csv")
normed_counts_long_dz <- normed_counts_dz_curves[,c(2:25,27)] %>% 
  #dplyr::filter(symbol == "TARDBP" | symbol == "UPF1" | symbol == "INSR") %>%
  pivot_longer(c(1:24)) %>%
  mutate(group = gsub("^[^_]+_|_[^_]+$", "", name))

normed_counts_long_dz %>%
  dplyr::filter(symbol == "TARDBP" | symbol == "STMN2" | symbol == "AARS" | symbol == "UNC13A" |
                  symbol == "SYNE1" | symbol == "CYFIP2") %>%
  ggplot(aes(x = group, y = value, color = symbol)) +
  geom_boxplot(alpha = 0.1, show.legend = F) +
  geom_point(position = position_dodge2(width = 1), show.legend = F) +
  #stat_pvalue_manual(stattest, hide.ns = T) + #[c(1,2),]) +
  #scale_y_log10() +
  facet_wrap(facets = vars(symbol), scales = "free_y") +
  scale_color_brewer(palette = "Set1") +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 90))
```

```{r}
#density plot
normed_counts_long_dz %>%
  ggplot(aes(x = value, color = group)) +
  geom_histogram(aes(y=..density..), colour="black", fill="white") +
  geom_density() +
  facet_wrap(facets = vars(group), nrow = 2) +
  scale_color_brewer(palette="Set1") +
  scale_x_log10() +
  labs(x = expression(paste("Lo", g[10], " counts")), y = "Density", color = "Group") +
  theme_classic()

mean_counts <- normed_counts_long_dz %>%
  group_by(group, symbol) %>%
  mutate(meanz = mean(value)) %>%
  mutate(varz = var(value))

mean_counts %>% 
  ggplot(aes(x=meanz, y=varz, color = group)) +
  geom_point() +
  geom_abline(intercept = 0, slope = 1, color="black") +
  facet_wrap(facets = vars(group), nrow = 2) +
  scale_color_brewer(palette="Set1") +
  scale_x_log10() +
  scale_y_log10() +
  theme_classic()

#quantile-quantile plot
normed_counts_long_dz %>%
  ggplot(aes(sample = value, color = group)) +
  geom_qq() + ###??
  geom_qq_line() + ###??
  facet_wrap(facets = vars(group), nrow = 2) +
  scale_color_brewer(palette="Set1") +
  #scale_x_log10() +
  theme_classic()

### pca analysis
pca_dz <- normed_counts_dz_curves[,c(1:25)] %>%
  #arrange(CTRL_ctrl_1) %>% ####can be better?
  distinct(Geneid, .keep_all = TRUE) %>%
  #select(c(1:)) %>%
  column_to_rownames('Geneid')
  #column_to_rownames('ensgene') #%>% t()

meta_df <- metadata_dir_dz %>% 
  #mutate(condition = factor(condition, levels = c(0,1))) %>%
  #mutate(treatment = factor(treatment, levels = c(0,1))) %>%
    #mutate(group = gsub("_[^_]+$", "", sample)) %>%
   mutate(level = as.character(level)) %>%
    column_to_rownames('sample') 

pp <- PCAtools::pca(pca_dz, metadata = meta_df, removeVar = 0.1)
screeplot(pp, axisLabSize = 18, titleLabSize = 22)
PCAtools::eigencorplot(pp,components = getComponents(pp),
                       metavars = c('condition',"level"))
#nice pca plot
biplot(pp,
       x = "PC1",
       y = "PC2",
       colby = 'level')
```


```{r}
colnames(featureCounts) <- c("ensgene", 
                           "DOX_0.0125_1", "DOX_0.0125_2", "DOX_0.0125_3",
                           "DOX_0.0187_1", "DOX_0.0187_2", "DOX_0.0187_3",
                           "DOX_0.021", "DOX_0.021_2", "DOX_0.021_3", 
                           "DOX_0.025_1", "DOX_0.025_2", "DOX_0.025_3", 
                           "DOX_0.075_1", "DOX_0.075_2", "DOX_0.075_3",
                           "NT_0_1", "NT_0_2", "NT_0_3", 
                           "gene_name")

pca_table <- featureCounts[,1:19] %>%
    column_to_rownames('ensgene')
annotation <- featureCounts[,c(1,20)] %>%
    column_to_rownames('ensgene')

path_to_meta_you_downloaded <-  "data/metadata.csv" #update this
meta_df <- fread(path_to_meta_you_downloaded)
meta_df <- meta_df %>%
    column_to_rownames('sample') 
meta_df[,1] <- as.factor(meta_df[,1])
meta_df[,2] <- as.factor(meta_df[,2])

pcaExplorer(countmatrix = pca_table, coldata = meta_df, annotation = annotation)


gene_list <-  c("SYT7", "RSF1", "KCNQ2", "HDGFL2", "AGRN", "ACTL6B", "MYO18A", "MYO1C", "KALRN", "DNM1", "CAMK2B", "ADGRB1","TARDBP")
grepper <- c("a", "b", "c", "f", "h")
names(grepper) <- c("dox00125", "dox00187", "dox0021", "dox0025", "dox0075")
for (j in grepper) {
my.dds <- run_standard_deseq("data/feature_count", #change here!!
                             base_grep = "NT",
                             contrast_grep = "DOX",  
                             grep_pattern = j,
                             baseName = "Untreated",
                             contrastName = 'TDP43KD')

bbb <- label_significant(my.dds$results_table, gene_list, log2FoldCut = 1.5, log10padj = 5) + annotate("text", x=4, y=-0.25, label=names(j))
plot(bbb)
}
```

splicing volcano and upset plot
```{r}
#base <- "Control"
contrast <- c("TDP43KD002", "TDP43KD003", "TDP43KD004", "TDP43KD005", "TDP43KD0075", "TDP43KD01", "TDP43KD1")
datalista <- list()

for (i in contrast) {
  input_bc_splicing <- fread(paste0("~/Documents/GitHub/tdp43_concentration/data/majiq_dz/Control-", i, "_annotated_junctions.csv"))
  names(input_bc_splicing)[12] <- "base_mean_psi"
  names(input_bc_splicing)[13] <- "contrast_mean_psi"
  assign(paste0("DS_",i), input_bc_splicing)
  a <- splicing_dots_tables_function(input_bc_splicing) + annotate("text", x=0.9, y=0.1, label=i)
  #plot(a)

  datalista[[i]] <- input_bc_splicing
}
big_delta_dz <- rbindlist(datalista, idcol = TRUE)
big_delta_dz$.id <- factor(big_delta_dz$.id, levels = contrast)



input_list <- c("DZ_curves_0","DZ_curves_002", "DZ_curves_003", "DZ_curves_004", "DZ_curves_005", "DZ_curves_0075", "DZ_curves_01", "DZ_curves_1")
datalist <- list()

for (i in input_list) {
  input_splicing <- fread(paste0("~/Documents/GitHub/tdp43_concentration/data/majiq_dz/", i, "_parsed.csv"))
  datalist[[i]] <- input_splicing
}
big_data <- rbindlist(datalist, idcol = TRUE)
big_data$.id <- factor(big_data$.id, levels = input_list)


###change to novel script
#spaghetti_plot(big_data)
#spaghetti_plot_normal(big_data)

```


```{r}
big_data_matrix_dz = big_delta_dz %>%
  filter(junc_cat != "annotated") %>%
  mutate(is_a_cryptic = base_mean_psi < 0.05 & mean_dpsi_per_lsv_junction > 0.1) %>%
           #abs(mean_dpsi_per_lsv_junction) > 0.15 & probability_changing > 0.9) %>% # 
  mutate(name = glue::glue("{gene_name}|{junc_cat}|{paste_into_igv_junction}")) %>%
  dplyr::select(name,is_a_cryptic,.id) %>%
  unique() %>%
  filter(is_a_cryptic == T)%>% 
  pivot_wider(values_from = "is_a_cryptic",
              id_cols = "name",
              names_from = ".id",
              values_fill = FALSE, values_fn = sum) %>%
  column_to_rownames('name') %>%
  mutate_all(as.logical)
  
upset(big_data_matrix_dz * 1, nsets = 7,
      order.by = "freq", keep.order = T, nintersects = 23,
      mainbar.y.label = "Junction Intersections", 
      sets.x.label = "Junction Per Level of TDP-43 KD")
```

```{r}
postar_bed <- fread("~/Desktop/rbp_bed/human_RBP_binding_sites_sorted.bed", 
                    col.names = c("seqnames", "start", "end", "dataset", "score", "strand", "QC")) ####use TDP only

ensembl <- biomaRt::useEnsembl(biomart = "genes", dataset = "hsapiens_gene_ensembl")
humandb <- biomaRt::getBM(attributes = c("external_gene_name", 
                                         "ensembl_gene_id", "ensembl_transcript_id", "transcript_appris", 
                                         "uniprotswissprot", "uniprotsptrembl", "uniprot_isoform"), 
                          mart = ensembl)
princ <- humandb[which(humandb$transcript_appris == "principal1" |
                       humandb$transcript_appris == "principal2" | 
                       humandb$transcript_appris == "principal3" | 
                       humandb$transcript_appris == "principal4" |
                       humandb$transcript_appris == "principal5"), ]


cds_regions = cdsBy(TxDb.Hsapiens.UCSC.hg38.knownGene, "tx",use.names = TRUE)
cds_regions = unlist(cds_regions)
cds_regions$transcript_id = gsub("\\..*", "", names(cds_regions))

exons_regions = exonsBy(TxDb.Hsapiens.UCSC.hg38.knownGene, "tx",use.names = TRUE)
exons_regions = unlist(exons_regions)
exons_regions$transcript_id = gsub("\\..*", "", names(exons_regions))

file_path = "data/sj_tabs/normalized_annotated/"
suffix = "_normalized_annotated.csv"
psi_files = list.files(file_path,
                       pattern = suffix,
                       full.names = TRUE)
psi_list_full = purrr::map(psi_files,my_clean_reader)
samp_ids = base::tolower(purrr::simplify(purrr::map(psi_files, basename)))
samp_ids = gsub(suffix,"",samp_ids)
##add on the name as an additional column
psi_list_full = purrr::map2(psi_list_full, samp_ids, ~cbind(.x, SampleID = .y))
psi_list_full = purrr::map(psi_list_full,my_name_fixer)
psi_list_full = data.table::rbindlist(psi_list_full)

for (i in contrast) {
 transcript_bind_plot("UNC13A", i)}
```

after running IRFinder, compute DE
```{r}
ir_top_level = "data/irfinder/"
experiment = make_deframe_irfinder(ir_top_level,base_grep = "NT",
                                   contrast_grep = "DOX")$metadata
paths = make_deframe_irfinder(ir_top_level,base_grep = "NT",
                                   contrast_grep = "DOX")$path

for (j in c("0.0125", "0.0187", "0.021", "0.025", "0.075")) {
    experiment <- filter(experiment, grepl(j, SampleNames) | Condition == "control")
    paths <- paths[grepl(j, paths) | grepl("NT", paths)]
}

dds = run_deseq_ir(paths,experiment = experiment)
result = return_formated_results(dds)
result0075 <- result %>% 
    mutate(ir_change = TDPKD -  control)

result_merged <- dplyr::bind_rows(list(result00125, result00187, result0021, result0025, result0075), .id = 'source')
```
