---
title: "TDP43 concentration dependent analysis"
author: "Matteo Zanovello"
date: "2022-01-10"
output:
---

```{r setup, include = False}
knitr::opts_chunk$set(warning = FALSE, message = FALSE)
if (!require("pacman")) {install.packages("pacman")}
if (!requireNamespace("BiocManager", quietly = TRUE)) {install.packages("BiocManager")}
pacman::p_load(tidyverse, data.table, janitor, ggpubr, ggrepel, magrittr, biomaRt, tximport, rlang,
               devtools, readxl, DESeq2, edgeR, annotables, rtracklayer, ensembldb,
               PCAtools, clusterProfiler, "org.Hs.eg.db", enrichplot, pcaExplorer, topGO, UpSetR,
               ggtranscript, "TxDb.Hsapiens.UCSC.hg38.knownGene", GenomicRanges)

source(here::here('scripts','salmon_deseq2.R'))            
source(here::here('scripts','make_volcano_plot.R'))
source(here::here('scripts','splicing_volcanoplot.R'))
#source(here::here('scripts','trial_graph_concentration.R'))

#source(here::here('scripts','run_de_irfinder.R'))
#source(here::here('scripts','ggtranscript.R'))
```

this organizes all the data from the many .bam files and creates one data frame you can work with,
then runs the make_deseq_dfs() function and run_standard_deseq() and eventually make graphs
```{r}
#update with new salmon quant
salmon_quant_directory_sy5y = "/Users/matteozanovello/Documents/Github/tdp43_concentration/data/salmon_curves/"
metadata_dir = fread("/Users/matteozanovello/Documents/Github/tdp43_concentration/data/metadata.csv")
levels_sy5y <- c(0.0125, 0.0187, 0.021, 0.025, 0.075)
salmon_deseq2(levels_sy5y, salmon_quant_directory_sy5y, metadata_dir)
assign("normed_counts_sy5y_curves", normed_counts)

#datalimbo <- list()
for (i in levels_sy5y) {
  salmon_deseq2(i, salmon_quant_directory_sy5y, metadata_dir)
  bbb <- label_significant(results_table, log2FoldCut = 2, log10padj = 5) + annotate("text", x=0.9, y=0.1, label=i) ###volcano plots here for simplicity - but would go later in the story
  #plot(bbb)
  assign(paste0("DE_volcano_",i),bbb)
  assign(paste0("results_table_sy5y_curves_",i),results_table)
  assign(paste0("normed_counts_sy5y_curves_",i), normed_counts)
}

AppendMe <- function(dfNames) {
  do.call(rbind, lapply(dfNames, function(x) {
    cbind(get(x), source = x)
  }))
}
list_de <- paste0("results_table_sy5y_curves_", levels_sy5y)
big_delta_de <- AppendMe(list_de) 
big_delta_de$source <- factor(big_delta_de$source, levels = list_de, labels = levels_sy5y)
list_gene <- c("UNC13A", "STMN2", "INSR"#, #"AARS", #"UNC13B", 
               #"CELF5", "ELAVL3", "INSR", "ATG4B", #"KCNQ2", #"RANBP1", 
               #"NUP188","SYT7", #"WASL",
                #"RAB1GAP", #"AGRN","PFKP", 
                 #"HDGFL2","ARHGAP32", "CEP290", "PRELID3A", "KALRN",
                 #"CAMK2B", "ADCY1", "TRAPPC12", "EPB41L4A", "IQCE", "SEMA4D"
              #   "CYFIP2", "SYNE1"
               )

write.table(big_delta_de, file = "~/Desktop/diff_expr_sh.csv", sep = ",", row.names = F) #"normed_counts_dz_all.csv")

plot_de <- big_delta_de %>%
  mutate(color_gene_name = ifelse(log2FoldChange > 0, "up","down")) %>%
  mutate(alpha_gene_name = ifelse(padj < 0.05, "sig", "non_sig")) %>%
  mutate(gene_label = case_when(symbol %in% list_gene & alpha_gene_name == "sig" ~ symbol, T ~ ""))
plot_de_meta <- as.data.frame(table(plot_de$color_gene_name, plot_de$alpha_gene_name, plot_de$source))

plot_de %>%
  ggplot(aes(x = log2FoldChange, y = source)) +
  geom_point(aes(color = color_gene_name, alpha = alpha_gene_name),
             size = 0.2,
             position = position_jitter(width = 0, height = 0.4),
             show.legend = F) +
  geom_text(data = plot_de_meta[plot_de_meta$Var2 == "sig",], 
            aes(x = ifelse(Var1 == "down", -8, 8), y = Var3, label = Freq, color = Var1),
            size = 5,
            show.legend = F) +
  geom_vline(xintercept = 0) +
  geom_text_repel(aes(label = gene_label), max.overlaps = Inf) +
  #coord_flip() +
  scale_color_brewer(palette = "Set1") +
  scale_alpha_manual(values = c(0.2, 1)) +
  xlim(c(-10,10)) +
  xlab('Gene expression Log2 Fold-Change compared to controls') +
  ylab('TDP-43 KD level') +
  scale_y_discrete(labels = c("+", "++", "+++", "++++", "+++++")) +
  #scale_x_continuous(trans = cubroot_trans) +
  theme_classic() +
  theme(text = element_text(size = 22))
#ggsave("~/Desktop/de_sy5y_with_dark.png", width = 12)

write.table(normed_counts_sy5y_curves, "~/Desktop/normed_counts_sh.csv", sep = ",",  row.names = F)
normed_counts_long <- normed_counts_sy5y_curves[,c(2:19,21)] %>% 
  #dplyr::filter(symbol == "TARDBP" | symbol == "UPF1" | symbol == "INSR") %>%
  pivot_longer(c(1:18)) %>%
  mutate(group = gsub("^[^_]+_|_[^_]+$", "", name))
###single gene plot
normed_counts_long %>%
  dplyr::filter(symbol == "TARDBP" | symbol == "STMN2" | symbol == "INSR" | symbol == "UNC13A") %>%
  ggplot(aes(x = group, y = value, color = symbol)) +
  geom_boxplot(alpha = 0.1, show.legend = F) +
  geom_point(position = position_dodge2(width = 1), show.legend = F) +
  #stat_pvalue_manual(stattest, hide.ns = T) + #[c(1,2),]) +
  #scale_y_log10() +
  facet_wrap(facets = vars(symbol), scales = "free_y") +
  scale_color_brewer(palette = "Set1") +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 90))
```

```{r}
normed_counts_tdp <- normed_counts_long %>% 
  filter(symbol == "TARDBP") 
normed_counts_tdp %>% 
  ggplot(aes(x = group, y = value/10305.2054 *100)) +
  geom_boxplot() +
  labs(x = "Doxycycline concentration", y = "Percentage of TDP-43 compared to controls") +
  theme_classic()

```

```{r}
#density plot
normed_counts_long <- normed_counts_sy5y_curves[,c(2:19,21)] %>% #change 
  #dplyr::filter(symbol == "TARDBP" | symbol == "UPF1" | symbol == "INSR") %>%
  pivot_longer(c(1:18)) %>%
  mutate(group = gsub("^[^_]+_|_[^_]+$", "", name))

normed_counts_long %>%
  ggplot(aes(x = value, color = group)) +
  geom_histogram(aes(y=..density..), colour="black", fill="white") +
  geom_density() +
  facet_wrap(facets = vars(group)) +
  scale_color_brewer(palette="Set1") +
  scale_x_log10() +
  labs(x = expression(paste("Lo", g[10], " counts")), y = "Density", color = "Group") +
  theme_classic()

mean_counts <- normed_counts_long %>%
  group_by(group, symbol) %>%
  mutate(meanz = mean(value)) %>%
  mutate(varz = var(value))

mean_counts %>% 
  ggplot(aes(x=meanz, y=varz, color = group)) +
  geom_point() +
  geom_abline(intercept = 0, slope = 1, color="black") +
  facet_wrap(facets = vars(group)) +
  scale_color_brewer(palette="Set1") +
  scale_x_log10() +
  scale_y_log10() +
  theme_classic()

#quantile-quantile plot
normed_counts_long %>%
  ggplot(aes(sample = value, color = group)) +
  geom_qq(distribution = qnbinom, dparams = c(size = 10000, prob = 0.90)) + ###??
  geom_qq_line(distribution = qnbinom, dparams = c(size = 10000, prob = 0.90)) + ###??
  facet_wrap(facets = vars(group)) +
  scale_color_brewer(palette="Set1") +
  #scale_x_log10() +
  theme_classic()

### pca analysis
pca <- normed_counts_sy5y_curves[,c(1:19)] %>%
  #arrange(CTRL_ctrl_1) %>% ####can be better?
  distinct(Geneid, .keep_all = TRUE) %>%
  #select(c(1:)) %>%
  column_to_rownames('Geneid')
  #column_to_rownames('ensgene') #%>% t()

meta_df <- metadata_dir %>% 
  #mutate(condition = factor(condition, levels = c(0,1))) %>%
  #mutate(treatment = factor(treatment, levels = c(0,1))) %>%
    #mutate(group = gsub("_[^_]+$", "", sample)) %>%
  #mutate(level = as.character(level)) %>%
    column_to_rownames('sample') 

pp <- PCAtools::pca(pca, metadata = meta_df, removeVar = 0.1)
screeplot(pp, axisLabSize = 18, titleLabSize = 22)
PCAtools::eigencorplot(pp,components = getComponents(pp),
                       metavars = c('condition',"level"))
#nice pca plot
biplot(pp,
       x = "PC1",
       y = "PC3",
       colby = 'level')
```

splicing volcano and upset plot
```{r}
#base <- "noDox"
contrast <- c("dox00125", "dox00187", "dox0021", "dox0025", "dox0075")
datalista <- list()
for (i in contrast) {
  input_splicing <- fread(paste0("/Users/matteozanovello/Documents/GitHub/tdp43_concentration/data/majiq_delta/noDox-", i, "_annotated_junctions.csv"))
  names(input_splicing)[12] <- "base_mean_psi"
  names(input_splicing)[13] <- "contrast_mean_psi"
  assign(paste0("DS_",i), input_splicing)

  a <- splicing_dots_tables_function(input_splicing) + annotate("text", x=0.9, y=0.1, label=i)
  #plot(a)

  datalista[[i]] <- input_splicing
}
big_delta <- rbindlist(datalista, idcol = TRUE)
big_delta$.id <- factor(big_delta$.id, levels = contrast)

input_list <- c("no_dox","dox_0125", "dox_0187", "dox_021", "dox_025", "dox_075")
datalist <- list()
for (i in input_list) {
  #input <- paste(i, "parsed", sep = "_")
  #input_csv <- paste(input, "csv", sep = ".")
  input_splicing <- fread(paste0("/Users/matteozanovello/Documents/GitHub/tdp43_concentration/data/majiq_single/", i, "_parsed.csv"))
  datalist[[i]] <- input_splicing
}
big_data <- rbindlist(datalist, idcol = TRUE)
big_data$.id <- factor(big_data$.id, levels = input_list)

###upset
big_data_matrix = big_delta %>%
  filter(junc_cat != "annotated") %>%
  filter(gene_name %in% list_gene) %>% 
  mutate(is_a_cryptic = base_mean_psi < 0.05 & mean_dpsi_per_lsv_junction > 0.1) %>%
           #abs(mean_dpsi_per_lsv_junction) > 0.15 & probability_changing > 0.9) %>% # 
  mutate(name = glue::glue("{gene_name}|{junc_cat}|{paste_into_igv_junction}")) %>%
  dplyr::select(name,is_a_cryptic,.id) %>%
  unique() %>%
  filter(is_a_cryptic == T)%>% 
  pivot_wider(values_from = "is_a_cryptic",
              id_cols = "name",
              names_from = ".id",
              values_fill = FALSE, values_fn = sum) %>%
  column_to_rownames('name') %>%
  mutate_all(as.logical)
  
upset(big_data_matrix * 1,
      order.by = "freq", keep.order = TRUE, nintersects = 22,
      mainbar.y.label = "Junction intersections", 
      sets.x.label = "Junction per TDP-43 KD level",
      text.scale = 2)

```

```{r}
#spaghetti - fix
spaghetti_plot(big_data) ###fix - use the new script?? - use new data cleaned
#spaghetti_plot_normal(big_data)
```

```{r}
#create input for GO - DE (do it also on DS)
carmm <- results_table_sy5y_curves_0.075
cm_PR_up <- carmm %>% dplyr::filter(log2FoldChange > 0 & padj < 0.05) 
cm_PR_up <- cm_PR_up[order(-cm_PR_up$log2FoldChange),]
cm_PR_down <- carmm %>% dplyr::filter(log2FoldChange < 0 & padj < 0.05)
cm_PR_down <- cm_PR_down[order(cm_PR_down$log2FoldChange),]

#GO analysis
go_cm_PR <- gprofiler2::gost(query = list(Upregulated = cm_PR_up$symbol,
                                          Downregulated = cm_PR_down$symbol), 
                             sources = c("GO:BP", "GO:CC", "GO:MF"), organism = 'hsapiens', 
                             ordered_query = T, evcodes = TRUE)

go_cm_PR$result %>%
  dplyr::filter(-log10(p_value) > 40) %>%
  arrange(source, desc(query), desc(p_value)) %>%
  mutate(term_code = paste0(term_name, " (", term_id, ")")) %>%
  mutate(term_code=factor(term_code, levels=unique(term_code))) %>%
  ggplot(aes(y = term_code, group = query, x = -log10(p_value), fill = query)) +
  geom_col(position = "dodge", width = 0.8) +
  #geom_point(size = 10, show.legend = F) +
  geom_text(aes(label = intersection_size), 
            nudge_x = -1, size = 0.8) +
  facet_grid(facets = vars(source), 
             scales = "free_y", space = "free_y") +
  theme_classic() +
  scale_fill_manual(values = c("#FAD510", "#3B9AB2")) +
  labs(title = "Enriched GO terms for significant genes in DOXvsCTRL", 
       y = "Term name and code", 
       x = expression(paste("-lo", g[10], " (", italic("P"), " value)")), fill = "") + 
  theme(legend.position = "top",
        legend.justification='left',
        legend.key.size = unit(0.5,"line"),
        strip.text = element_text(
          size = 4)) +
  theme(text = element_text(size = 6)) +
  theme(plot.title = element_text(hjust = -0.5, size = 8)) +
  theme(legend.text=element_text(size=4))


###GO on DS
carm_ds <- big_delta %>% filter(.id == "dox0025" & base_mean_psi < 0.05 & mean_dpsi_per_lsv_junction > 0.1) %>%
  arrange(-mean_dpsi_per_lsv_junction) %>% distinct(gene_name)
go_cm_DS <- gprofiler2::gost(query = list(Differentially_spliced = carm_ds$gene_name),
                                          #Downregulated = cm_DS_down$gene_name), 
                             sources = c("GO:BP", "GO:CC", "GO:MF"), organism = 'hsapiens', 
                             ordered_query = T, evcodes = TRUE)
go_cm_DS$result %>%
  #dplyr::filter(-log10(p_value) > 2) %>%
  arrange(source, desc(query), desc(p_value)) %>%
  mutate(term_code = paste0(term_name, " (", term_id, ")")) %>%
  mutate(term_code=factor(term_code, levels=unique(term_code))) %>%
  ggplot(aes(y = term_code, group = query, x = -log10(p_value), fill = query)) +
  geom_col(position = "dodge", width = 0.8) +
  #geom_point(size = 10, show.legend = F) +
  geom_text(aes(label = intersection_size), 
            nudge_x = -1, size = 0.8) +
  facet_grid(facets = vars(source), 
             scales = "free_y", space = "free_y") +
  theme_classic() +
  scale_fill_manual(values = c("#FAD510", "#3B9AB2")) +
  labs(title = "Enriched GO terms for significant genes in DOXvsCTRL", 
       y = "Term name and code", 
       x = expression(paste("-lo", g[10], " (", italic("P"), " value)")), fill = "") + 
  theme(legend.position = "top",
        legend.justification='left',
        legend.key.size = unit(0.5,"line"),
        strip.text = element_text(
          size = 4)) +
  theme(text = element_text(size = 6)) +
  theme(plot.title = element_text(hjust = -0.5, size = 8)) +
  theme(legend.text=element_text(size=4))

###term overlap splicing and expression
```

##ggtranscript
```{r}
postar_bed <- fread("~/Desktop/rbp_bed/human_RBP_binding_sites_sorted.bed", 
                    col.names = c("seqnames", "start", "end", "dataset", "score", "strand", "QC")) ####use TDP only

ensembl <- biomaRt::useEnsembl(biomart = "genes", dataset = "hsapiens_gene_ensembl")
humandb <- biomaRt::getBM(attributes = c("external_gene_name", 
                                         "ensembl_gene_id", "ensembl_transcript_id", "transcript_appris", 
                                         "uniprotswissprot", "uniprotsptrembl", "uniprot_isoform"), 
                          mart = ensembl)
princ <- humandb[which(humandb$transcript_appris == "principal1" |
                       humandb$transcript_appris == "principal2" | 
                       humandb$transcript_appris == "principal3" | 
                       humandb$transcript_appris == "principal4" |
                       humandb$transcript_appris == "principal5"), ]


cds_regions = cdsBy(TxDb.Hsapiens.UCSC.hg38.knownGene, "tx",use.names = TRUE)
cds_regions = unlist(cds_regions)
cds_regions$transcript_id = gsub("\\..*", "", names(cds_regions))

exons_regions = exonsBy(TxDb.Hsapiens.UCSC.hg38.knownGene, "tx",use.names = TRUE)
exons_regions = unlist(exons_regions)
exons_regions$transcript_id = gsub("\\..*", "", names(exons_regions))

for (i in contrast) {
 transcript_bind_plot("UNC13A", i) ###needs edit when calling CE table
  }
```

after running IRFinder, compute DE
```{r}
#ir_top_level = "data/irfinder/"
#experiment = make_deframe_irfinder(ir_top_level,base_grep = "NT",
#                                   contrast_grep = "DOX")$metadata
#paths = make_deframe_irfinder(ir_top_level,base_grep = "NT",
#                                   contrast_grep = "DOX")$path

#for (j in c("0.0125", "0.0187", "0.021", "0.025", "0.075")) {
#    experiment <- filter(experiment, grepl(j, SampleNames) | Condition == "control")
#    paths <- paths[grepl(j, paths) | grepl("NT", paths)]
#}

#dds = run_deseq_ir(paths,experiment = experiment)
#result = return_formated_results(dds)
#result0075 <- result %>% 
#    mutate(ir_change = TDPKD -  control)

#result_merged <- dplyr::bind_rows(list(result00125, result00187, result0021, result0025, result0075), .id = 'source')
```
